# Stage 2 (Data Pre-Processing)

Anggota Team 4:
*   Akhmad Yuzfa Salvian Idris
*   Arief Rahman Hakim
*   Bernardus Valentino
*   Milenia Nadia Afifah Puspitasari
*   Sean Frederic Wijaya

## EDA Conclusions

*   Terdapat **missing values**, yakni kolom Age, TypeofContact, DurationOfPitch, NumberOfFollowups, PreferredPropertyStar, NumberOfTrips, NumberOfChildrenVisiting, dan MonthlyIncome sehingga harus diurus ketika preprocessing
*  Feature-feature yang **seharusnya merupakan kategorik** yakni ProdTaken, CityTier, PreferredPropertyStar, Passport, PitchSatisfactionScore, dan OwnCar sehingga harus diubah saat preprocessing
*   Distribusi nilai di kolom Age, NumberOfPersonVisiting, NumberOfFollowups, NumberOfTrips, NumberOfChildrenVisiting, dan MonthlyIncome terlihat **hampir normal** sedangkan dari boxplot dan histplot, feature DurationOfPitch, NumberOfTrips, dan MonthlyIncome terlihat **right skewed** sehingga harus distandardisasi saat preprocessing
*   **Jumlah unik dari setiap feature cukup sedikit**. Ada kemungkinan akan berguna untuk klasifikasi
*   Feature **Gender memiliki jumlah unik 3**, sehingga perlu ditinjau lebih lanjut
*   Feature yang **tidak memiliki outlier** yakni Age dan NumberOfChildrenVisiting, lainnya memiliki outlier
*   Seluruh feature memiliki korelasi lemah dengan target
*   **Top 3 feature yang berkorelasi kuat dengan target**: Passport, Age, dan MonthlyIncome
*   Feature OwnCar, NumberOfChildrenVisiting, dan NumberOfPersonVisiting, memiliki **korelasi yang paling lemah dengan target** sehingga mungkin nantinya fitur-fitur ini dapat diabaikan
*   **Feature yang saling berkorelasi kuat/redundan (>0.3)**: NumberOfPersonVisiting dengan NumberOfChildrenVisiting, NumberOfPersonVisiting dengan NumberOfFollowups, dan Age dengan MonthlyIncome. Nantinya salah satu dari variabel yang berkorelasi akan didrop (**akan dilihat dari nilai VIF (Variance Inflation Factor) nya**).
*   **Kemungkinan feature yang didrop**: OwnCar, NumberOfChildrenVisiting, NumberOfPersonVisiting, dan MonthlyIncome

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('Travel.csv')
df

CustomerID  ProdTaken   Age    TypeofContact  CityTier  DurationOfPitch  \
0         200000          1  41.0     Self Enquiry         3              6.0   
1         200001          0  49.0  Company Invited         1             14.0   
2         200002          1  37.0     Self Enquiry         1              8.0   
3         200003          0  33.0  Company Invited         1              9.0   
4         200004          0   NaN     Self Enquiry         1              8.0   
...          ...        ...   ...              ...       ...              ...   
4883      204883          1  49.0     Self Enquiry         3              9.0   
4884      204884          1  28.0  Company Invited         1             31.0   
4885      204885          1  52.0     Self Enquiry         3             17.0   
4886      204886          1  19.0     Self Enquiry         3             16.0   
4887      204887          1  36.0     Self Enquiry         1             14.0   

          Occupation  Gender  NumberOfPersonVisiting  NumberOfFollowups  \
0           Salaried  Female                       3                3.0   
1           Salaried    Male                       3                4.0   
2        Free Lancer    Male                       3                4.0   
3           Salaried  Female                       2                3.0   
4     Small Business    Male                       2                3.0   
...              ...     ...                     ...                ...   
4883  Small Business    Male                       3                5.0   
4884        Salaried    Male                       4                5.0   
4885        Salaried  Female                       4                4.0   
4886  Small Business    Male                       3                4.0   
4887        Salaried    Male                       4                4.0   

     ProductPitched  PreferredPropertyStar MaritalStatus  NumberOfTrips  \
0            Deluxe                    3.0        Single            1.0   
1            Deluxe                    4.0      Divorced            2.0   
2             Basic                    3.0        Single            7.0   
3             Basic                    3.0      Divorced            2.0   
4             Basic                    4.0      Divorced            1.0   
...             ...                    ...           ...            ...   
4883         Deluxe                    4.0     Unmarried            2.0   
4884          Basic                    3.0        Single            3.0   
4885       Standard                    4.0       Married            7.0   
4886          Basic                    3.0        Single            3.0   
4887          Basic                    4.0     Unmarried            3.0   

      Passport  PitchSatisfactionScore  OwnCar  NumberOfChildrenVisiting  \
0            1                       2       1                       0.0   
1            0                       3       1                       2.0   
2            1                       3       0                       0.0   
3            1                       5       1                       1.0   
4            0                       5       1                       0.0   
...        ...                     ...     ...                       ...   
4883         1                       1       1                       1.0   
4884         1                       3       1                       2.0   
4885         0                       1       1                       3.0   
4886         0                       5       0                       2.0   
4887         1                       3       1                       2.0   

         Designation  MonthlyIncome  
0            Manager        20993.0  
1            Manager        20130.0  
2          Executive        17090.0  
3          Executive        17909.0  
4          Executive        18468.0  
...              ...            ...  
4883         Manager        26576.0  
4884       Exe

## Merubah Tipe Data

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4888 entries, 0 to 4887
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CustomerID                4888 non-null   int64  
 1   ProdTaken                 4888 non-null   int64  
 2   Age                       4662 non-null   float64
 3   TypeofContact             4863 non-null   object 
 4   CityTier                  4888 non-null   int64  
 5   DurationOfPitch           4637 non-null   float64
 6   Occupation                4888 non-null   object 
 7   Gender                    4888 non-null   object 
 8   NumberOfPersonVisiting    4888 non-null   int64  
 9   NumberOfFollowups         4843 non-null   float64
 10  ProductPitched            4888 non-null   object 
 11  PreferredPropertyStar     4862 non-null   float64
 12  MaritalStatus             4888 non-null   object 
 13  NumberOfTrips             4748 non-null   float64
 14  Passport

In [4]:
df['ProdTaken'] = df['ProdTaken'].astype('str')
df['CityTier'] = df['CityTier'].astype('str')
df['PreferredPropertyStar'] = df['PreferredPropertyStar'].astype('str')
df['Passport'] = df['Passport'].astype('str')
df['PitchSatisfactionScore'] = df['PitchSatisfactionScore'].astype('str')
df['OwnCar'] = df['OwnCar'].astype('str')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4888 entries, 0 to 4887
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CustomerID                4888 non-null   int64  
 1   ProdTaken                 4888 non-null   object 
 2   Age                       4662 non-null   float64
 3   TypeofContact             4863 non-null   object 
 4   CityTier                  4888 non-null   object 
 5   DurationOfPitch           4637 non-null   float64
 6   Occupation                4888 non-null   object 
 7   Gender                    4888 non-null   object 
 8   NumberOfPersonVisiting    4888 non-null   int64  
 9   NumberOfFollowups         4843 non-null   float64
 10  ProductPitched            4888 non-null   object 
 11  PreferredPropertyStar     4888 non-null   object 
 12  MaritalStatus             4888 non-null   object 
 13  NumberOfTrips             4748 non-null   float64
 14  Passport

In [6]:
cats = ['TypeofContact','Occupation','Gender','ProductPitched','MaritalStatus','Designation', 'ProdTaken', 'CityTier', 'PreferredPropertyStar', 'Passport','PitchSatisfactionScore','OwnCar']
nums = ['CustomerID','Age','DurationOfPitch','NumberOfPersonVisiting','NumberOfFollowups','NumberOfTrips','NumberOfChildrenVisiting','MonthlyIncome']

In [7]:
df[nums].describe()

CustomerID          Age  DurationOfPitch  NumberOfPersonVisiting  \
count    4888.000000  4662.000000      4637.000000             4888.000000   
mean   202443.500000    37.622265        15.490835                2.905074   
std      1411.188388     9.316387         8.519643                0.724891   
min    200000.000000    18.000000         5.000000                1.000000   
25%    201221.750000    31.000000         9.000000                2.000000   
50%    202443.500000    36.000000        13.000000                3.000000   
75%    203665.250000    44.000000        20.000000                3.000000   
max    204887.000000    61.000000       127.000000                5.000000   

       NumberOfFollowups  NumberOfTrips  NumberOfChildrenVisiting  \
count        4843.000000    4748.000000               4822.000000   
mean            3.708445       3.236521                  1.187267   
std             1.002509       1.849019                  0.857861   
min             1.000000       1.000000                  0.000000   
25%             3.000000       2.000000                  1.000000   
50%             4.000000       3.000000                  1.000000   
75%             4.000000       4.000000                  2.000000   
max             6.000000      22.000000                  3.000000   

       MonthlyIncome  
count    4655.000000  
mean    23619.853491  
std      5380.698361  
min      1000.000000  
25%     20346.000000  
50%     22347.000000  
75%     25571.000000  
max     98678.000000

In [8]:
df[cats].describe()

TypeofContact Occupation Gender ProductPitched MaritalStatus  \
count           4863       4888   4888           4888          4888   
unique             2          4      3              5             4   
top     Self Enquiry   Salaried   Male          Basic       Married   
freq            3444       2368   2916           1842          2340   

       Designation ProdTaken CityTier PreferredPropertyStar Passport  \
count         4888      4888     4888                  4888     4888   
unique           5         2        3                     4        2   
top      Executive         0        1                   3.0        0   
freq          1842      3968     3190                  2993     3466   

       PitchSatisfactionScore OwnCar  
count                    4888   4888  
unique                      5      2  
top                         3      1  
freq                     1478   3032

## Memperbaiki Fitur Gender

In [9]:
df.replace(to_replace=r'Fe Male', value='Female', regex=True, inplace=True )

In [10]:
df[cats].describe()

TypeofContact Occupation Gender ProductPitched MaritalStatus  \
count           4863       4888   4888           4888          4888   
unique             2          4      2              5             4   
top     Self Enquiry   Salaried   Male          Basic       Married   
freq            3444       2368   2916           1842          2340   

       Designation ProdTaken CityTier PreferredPropertyStar Passport  \
count         4888      4888     4888                  4888     4888   
unique           5         2        3                     4        2   
top      Executive         0        1                   3.0        0   
freq          1842      3968     3190                  2993     3466   

       PitchSatisfactionScore OwnCar  
count                    4888   4888  
unique                      5      2  
top                         3      1  
freq                     1478   3032

## Missing Value

In [11]:
df.isna().sum()

CustomerID                    0
ProdTaken                     0
Age                         226
TypeofContact                25
CityTier                      0
DurationOfPitch             251
Occupation                    0
Gender                        0
NumberOfPersonVisiting        0
NumberOfFollowups            45
ProductPitched                0
PreferredPropertyStar         0
MaritalStatus                 0
NumberOfTrips               140
Passport                      0
PitchSatisfactionScore        0
OwnCar                        0
NumberOfChildrenVisiting     66
Designation                   0
MonthlyIncome               233
dtype: int64

In [12]:
df.head()

CustomerID ProdTaken   Age    TypeofContact CityTier  DurationOfPitch  \
0      200000         1  41.0     Self Enquiry        3              6.0   
1      200001         0  49.0  Company Invited        1             14.0   
2      200002         1  37.0     Self Enquiry        1              8.0   
3      200003         0  33.0  Company Invited        1              9.0   
4      200004         0   NaN     Self Enquiry        1              8.0   

       Occupation  Gender  NumberOfPersonVisiting  NumberOfFollowups  \
0        Salaried  Female                       3                3.0   
1        Salaried    Male                       3                4.0   
2     Free Lancer    Male                       3                4.0   
3        Salaried  Female                       2                3.0   
4  Small Business    Male                       2                3.0   

  ProductPitched PreferredPropertyStar MaritalStatus  NumberOfTrips Passport  \
0         Deluxe                   3.0        Single            1.0        1   
1         Deluxe                   4.0      Divorced            2.0        0   
2          Basic                   3.0        Single            7.0        1   
3          Basic                   3.0      Divorced            2.0        1   
4          Basic                   4.0      Divorced            1.0        0   

  PitchSatisfactionScore OwnCar  NumberOfChildrenVisiting Designation  \
0                      2      1                       0.0     Manager   
1                      3      1                       2.0     Manager   
2                      3      0                       0.0   Executive   
3                      5      1                       1.0   Executive   
4                      5      1                       0.0   Executive   

   MonthlyIncome  
0        20993.0  
1        20130.0  
2        17090.0  
3        17909.0  
4        18468.0

In [13]:
df.dropna(inplace=True, subset=['TypeofContact','NumberOfChildrenVisiting'])
df.isna().sum()

CustomerID                    0
ProdTaken                     0
Age                         226
TypeofContact                 0
CityTier                      0
DurationOfPitch             226
Occupation                    0
Gender                        0
NumberOfPersonVisiting        0
NumberOfFollowups            45
ProductPitched                0
PreferredPropertyStar         0
MaritalStatus                 0
NumberOfTrips               140
Passport                      0
PitchSatisfactionScore        0
OwnCar                        0
NumberOfChildrenVisiting      0
Designation                   0
MonthlyIncome               208
dtype: int64

In [14]:
df['DurationOfPitch'].fillna(df['DurationOfPitch'].median(), inplace=True)
df['NumberOfFollowups'].fillna(df['NumberOfFollowups'].median(), inplace=True)
df['MonthlyIncome'].fillna(df['MonthlyIncome'].median(), inplace=True)

In [15]:
df['Age'].fillna(df['Age'].mode()[0], inplace=True)
df['NumberOfTrips'].fillna(df['NumberOfTrips'].mode()[0], inplace=True)

In [16]:
df.isna().sum()

CustomerID                  0
ProdTaken                   0
Age                         0
TypeofContact               0
CityTier                    0
DurationOfPitch             0
Occupation                  0
Gender                      0
NumberOfPersonVisiting      0
NumberOfFollowups           0
ProductPitched              0
PreferredPropertyStar       0
MaritalStatus               0
NumberOfTrips               0
Passport                    0
PitchSatisfactionScore      0
OwnCar                      0
NumberOfChildrenVisiting    0
Designation                 0
MonthlyIncome               0
dtype: int64

## Duplicated Data

In [17]:
df.duplicated().sum()

0

## Outliers

Menggunakan IQR

Menggunakan Z-score

In [18]:
from scipy import stats
print(f'Jumlah baris sebelum memfilter outlier: {len(df)}')

filtered_entries = np.array([True] * len(df))

for col in ['DurationOfPitch','NumberOfPersonVisiting','NumberOfFollowups','NumberOfTrips','MonthlyIncome']:
    zscore = abs(stats.zscore(df[col])) # hitung absolute z-scorenya
    filtered_entries = (zscore < 3) & filtered_entries # keep yang kurang dari 3 absolute z-scorenya
    
df = df[filtered_entries] # filter, cuma ambil yang z-scorenya dibawah 3

print(f'Jumlah baris setelah memfilter outlier: {len(df)}')

Jumlah baris sebelum memfilter outlier: 4797
Jumlah baris setelah memfilter outlier: 4787


Digunakan Z-score karena terlalu banyak data yang harus didrop jika menggunakan kriteria outlier IQR 

## Normalization/Standardization

In [19]:
df.describe()

CustomerID          Age  DurationOfPitch  NumberOfPersonVisiting  \
count    4787.000000  4787.000000      4787.000000             4787.000000   
mean   202445.630457    37.396490        15.337790                2.906204   
std      1412.764584     9.066171         8.017699                0.725755   
min    200000.000000    18.000000         5.000000                1.000000   
25%    201225.500000    31.000000         9.000000                2.000000   
50%    202446.000000    36.000000        13.000000                3.000000   
75%    203673.500000    43.000000        19.000000                3.000000   
max    204887.000000    61.000000        36.000000                5.000000   

       NumberOfFollowups  NumberOfTrips  NumberOfChildrenVisiting  \
count        4787.000000    4787.000000               4787.000000   
mean            3.710257       3.176729                  1.190934   
std             0.998026       1.765771                  0.858025   
min             1.000000       1.000000                  0.000000   
25%             3.000000       2.000000                  1.000000   
50%             4.000000       3.000000                  1.000000   
75%             4.000000       4.000000                  2.000000   
max             6.000000       8.000000                  3.000000   

       MonthlyIncome  
count    4787.000000  
mean    23382.838730  
std      4892.611455  
min     16009.000000  
25%     20447.500000  
50%     22222.000000  
75%     25274.000000  
max     38677.000000

- min max scaler tidak mengubah distribusi data
- boleh setelah min max lanjut standardscaler
- transformasi log

In [20]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler

df['NumberOfTrips_std'] = StandardScaler().fit_transform(df['NumberOfTrips'].values.reshape(len(df), 1))
df['MonthlyIncome_std'] = StandardScaler().fit_transform(df['MonthlyIncome'].values.reshape(len(df), 1))
df['DurationOfPitch_std'] = StandardScaler().fit_transform(df['DurationOfPitch'].values.reshape(len(df), 1))

<ipython-input-20-80aedaf7c60c>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['NumberOfTrips_std'] = StandardScaler().fit_transform(df['NumberOfTrips'].values.reshape(len(df), 1))
<ipython-input-20-80aedaf7c60c>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['MonthlyIncome_std'] = StandardScaler().fit_transform(df['MonthlyIncome'].values.reshape(len(df), 1))
<ipython-input-20-80aedaf7c60c>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .lo

In [21]:
df.describe()

CustomerID          Age  DurationOfPitch  NumberOfPersonVisiting  \
count    4787.000000  4787.000000      4787.000000             4787.000000   
mean   202445.630457    37.396490        15.337790                2.906204   
std      1412.764584     9.066171         8.017699                0.725755   
min    200000.000000    18.000000         5.000000                1.000000   
25%    201225.500000    31.000000         9.000000                2.000000   
50%    202446.000000    36.000000        13.000000                3.000000   
75%    203673.500000    43.000000        19.000000                3.000000   
max    204887.000000    61.000000        36.000000                5.000000   

       NumberOfFollowups  NumberOfTrips  NumberOfChildrenVisiting  \
count        4787.000000    4787.000000               4787.000000   
mean            3.710257       3.176729                  1.190934   
std             0.998026       1.765771                  0.858025   
min             1.000000       1.000000                  0.000000   
25%             3.000000       2.000000                  1.000000   
50%             4.000000       3.000000                  1.000000   
75%             4.000000       4.000000                  2.000000   
max             6.000000       8.000000                  3.000000   

       MonthlyIncome  NumberOfTrips_std  MonthlyIncome_std  \
count    4787.000000       4.787000e+03       4.787000e+03   
mean    23382.838730      -1.956562e-15      -3.988581e-16   
std      4892.611455       1.000104e+00       1.000104e+00   
min     16009.000000      -1.232865e+00      -1.507295e+00   
25%     20447.500000      -6.664805e-01      -6.000160e-01   
50%     22222.000000      -1.000963e-01      -2.372884e-01   
75%     25274.000000       4.662879e-01       3.865745e-01   
max     38677.000000       2.731825e+00       3.126298e+00   

       DurationOfPitch_std  
count         4.787000e+03  
mean         -1.974954e-16  
std           1.000104e+00  
min          -1.289506e+00  
25%          -7.905575e-01  
50%          -2.916091e-01  
75%           4.568135e-01  
max           2.577344e+00

In [22]:
df.drop(columns=['DurationOfPitch','NumberOfTrips','MonthlyIncome'], inplace=True)
df.head()

C:\Users\ASUS\anaconda3\lib\site-packages\pandas\core\frame.py:4906: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  return super().drop(


CustomerID ProdTaken   Age    TypeofContact CityTier      Occupation  \
0      200000         1  41.0     Self Enquiry        3        Salaried   
1      200001         0  49.0  Company Invited        1        Salaried   
2      200002         1  37.0     Self Enquiry        1     Free Lancer   
3      200003         0  33.0  Company Invited        1        Salaried   
4      200004         0  35.0     Self Enquiry        1  Small Business   

   Gender  NumberOfPersonVisiting  NumberOfFollowups ProductPitched  \
0  Female                       3                3.0         Deluxe   
1    Male                       3                4.0         Deluxe   
2    Male                       3                4.0          Basic   
3  Female                       2                3.0          Basic   
4    Male                       2                3.0          Basic   

  PreferredPropertyStar MaritalStatus Passport PitchSatisfactionScore OwnCar  \
0                   3.0        Single        1                      2      1   
1                   4.0      Divorced        0                      3      1   
2                   3.0        Single        1                      3      0   
3                   3.0      Divorced        1                      5      1   
4                   4.0      Divorced        0                      5      1   

   NumberOfChildrenVisiting Designation  NumberOfTrips_std  MonthlyIncome_std  \
0                       0.0     Manager          -1.232865          -0.488510   
1                       2.0     Manager          -0.666480          -0.664917   
2                       0.0   Executive           2.165440          -1.286327   
3                       1.0   Executive          -0.666480          -1.118914   
4                       0.0   Executive          -1.232865          -1.004648   

   DurationOfPitch_std  
0            -1.164769  
1            -0.166872  
2            -0.915295  
3            -0.790558  
4            -0.915295

In [23]:
df.describe()

CustomerID          Age  NumberOfPersonVisiting  NumberOfFollowups  \
count    4787.000000  4787.000000             4787.000000        4787.000000   
mean   202445.630457    37.396490                2.906204           3.710257   
std      1412.764584     9.066171                0.725755           0.998026   
min    200000.000000    18.000000                1.000000           1.000000   
25%    201225.500000    31.000000                2.000000           3.000000   
50%    202446.000000    36.000000                3.000000           4.000000   
75%    203673.500000    43.000000                3.000000           4.000000   
max    204887.000000    61.000000                5.000000           6.000000   

       NumberOfChildrenVisiting  NumberOfTrips_std  MonthlyIncome_std  \
count               4787.000000       4.787000e+03       4.787000e+03   
mean                   1.190934      -1.956562e-15      -3.988581e-16   
std                    0.858025       1.000104e+00       1.000104e+00   
min                    0.000000      -1.232865e+00      -1.507295e+00   
25%                    1.000000      -6.664805e-01      -6.000160e-01   
50%                    1.000000      -1.000963e-01      -2.372884e-01   
75%                    2.000000       4.662879e-01       3.865745e-01   
max                    3.000000       2.731825e+00       3.126298e+00   

       DurationOfPitch_std  
count         4.787000e+03  
mean         -1.974954e-16  
std           1.000104e+00  
min          -1.289506e+00  
25%          -7.905575e-01  
50%          -2.916091e-01  
75%           4.568135e-01  
max           2.577344e+00

## Feature encoding

In [24]:
df[cats].describe()

TypeofContact Occupation Gender ProductPitched MaritalStatus  \
count           4787       4787   4787           4787          4787   
unique             2          4      2              5             4   
top     Self Enquiry   Salaried   Male          Basic       Married   
freq            3383       2317   2869           1825          2294   

       Designation ProdTaken CityTier PreferredPropertyStar Passport  \
count         4787      4787     4787                  4787     4787   
unique           5         2        3                     4        2   
top      Executive         0        1                   3.0        0   
freq          1825      3879     3124                  2937     3389   

       PitchSatisfactionScore OwnCar  
count                    4787   4787  
unique                      5      2  
top                         3      1  
freq                     1445   2961

### Label Encoding

In [25]:
df['TypeofContact_label'] = df['TypeofContact'].astype('category').cat.codes
df['Passport_label'] = df['Passport'].astype('category').cat.codes
df['OwnCar_label'] = df['OwnCar'].astype('category').cat.codes
df['Gender_label'] = df['Gender'].astype('category').cat.codes
df['PreferredPropertyStar_label'] = df['PreferredPropertyStar'].astype('category').cat.codes
df['PitchSatisfactionScore_label'] = df['PitchSatisfactionScore'].astype('category').cat.codes
df['CityTier_label'] = df['CityTier'].astype('category').cat.codes

<ipython-input-25-86ed11dab95e>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['TypeofContact_label'] = df['TypeofContact'].astype('category').cat.codes
<ipython-input-25-86ed11dab95e>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Passport_label'] = df['Passport'].astype('category').cat.codes
<ipython-input-25-86ed11dab95e>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the

In [26]:
df.head()

CustomerID ProdTaken   Age    TypeofContact CityTier      Occupation  \
0      200000         1  41.0     Self Enquiry        3        Salaried   
1      200001         0  49.0  Company Invited        1        Salaried   
2      200002         1  37.0     Self Enquiry        1     Free Lancer   
3      200003         0  33.0  Company Invited        1        Salaried   
4      200004         0  35.0     Self Enquiry        1  Small Business   

   Gender  NumberOfPersonVisiting  NumberOfFollowups ProductPitched  ...  \
0  Female                       3                3.0         Deluxe  ...   
1    Male                       3                4.0         Deluxe  ...   
2    Male                       3                4.0          Basic  ...   
3  Female                       2                3.0          Basic  ...   
4    Male                       2                3.0          Basic  ...   

  NumberOfTrips_std MonthlyIncome_std DurationOfPitch_std TypeofContact_label  \
0         -1.232865         -0.488510           -1.164769                   1   
1         -0.666480         -0.664917           -0.166872                   0   
2          2.165440         -1.286327           -0.915295                   1   
3         -0.666480         -1.118914           -0.790558                   0   
4         -1.232865         -1.004648           -0.915295                   1   

  Passport_label  OwnCar_label Gender_label  PreferredPropertyStar_label  \
0              1             1            0                            0   
1              0             1            1                            1   
2              1             0            1                            0   
3              1             1            0                            0   
4              0             1            1                            1   

   PitchSatisfactionScore_label  CityTier_label  
0                             1               2  
1                             2               0  
2                             2               0  
3                             4               0  
4                             4               0  

[5 rows x 27 columns]

### One Hot Encoding

In [27]:
col = ['Occupation','ProductPitched','MaritalStatus','Designation']
for cat in col:
    onehots = pd.get_dummies(df[cat], prefix=cat)
    df = df.join(onehots)
    
df.head()

CustomerID ProdTaken   Age    TypeofContact CityTier      Occupation  \
0      200000         1  41.0     Self Enquiry        3        Salaried   
1      200001         0  49.0  Company Invited        1        Salaried   
2      200002         1  37.0     Self Enquiry        1     Free Lancer   
3      200003         0  33.0  Company Invited        1        Salaried   
4      200004         0  35.0     Self Enquiry        1  Small Business   

   Gender  NumberOfPersonVisiting  NumberOfFollowups ProductPitched  ...  \
0  Female                       3                3.0         Deluxe  ...   
1    Male                       3                4.0         Deluxe  ...   
2    Male                       3                4.0          Basic  ...   
3  Female                       2                3.0          Basic  ...   
4    Male                       2                3.0          Basic  ...   

  ProductPitched_Super Deluxe MaritalStatus_Divorced MaritalStatus_Married  \
0                           0                      0                     0   
1                           0                      1                     0   
2                           0                      0                     0   
3                           0                      1                     0   
4                           0                      1                     0   

  MaritalStatus_Single MaritalStatus_Unmarried  Designation_AVP  \
0                    1                       0                0   
1                    0                       0                0   
2                    1                       0                0   
3                    0                       0                0   
4                    0                       0                0   

  Designation_Executive  Designation_Manager  Designation_Senior Manager  \
0                     0                    1                           0   
1                     0                    1                           0   
2                     1                    0                           0   
3                     1                    0                           0   
4                     1                    0                           0   

   Designation_VP  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 45 columns]

In [28]:
df.drop(columns=['Occupation','Gender','ProductPitched','MaritalStatus','Designation', 'CityTier', 
                 'PreferredPropertyStar', 'PitchSatisfactionScore', 'TypeofContact', 'Passport', 'OwnCar'], inplace=True)
df

CustomerID ProdTaken   Age  NumberOfPersonVisiting  NumberOfFollowups  \
0         200000         1  41.0                       3                3.0   
1         200001         0  49.0                       3                4.0   
2         200002         1  37.0                       3                4.0   
3         200003         0  33.0                       2                3.0   
4         200004         0  35.0                       2                3.0   
...          ...       ...   ...                     ...                ...   
4883      204883         1  49.0                       3                5.0   
4884      204884         1  28.0                       4                5.0   
4885      204885         1  52.0                       4                4.0   
4886      204886         1  19.0                       3                4.0   
4887      204887         1  36.0                       4                4.0   

      NumberOfChildrenVisiting  NumberOfTrips_std  MonthlyIncome_std  \
0                          0.0          -1.232865          -0.488510   
1                          2.0          -0.666480          -0.664917   
2                          0.0           2.165440          -1.286327   
3                          1.0          -0.666480          -1.118914   
4                          0.0          -1.232865          -1.004648   
...                        ...                ...                ...   
4883                       1.0          -0.666480           0.652718   
4884                       2.0          -0.100096          -0.443744   
4885                       3.0           2.165440           1.724650   
4886                       2.0          -0.100096          -0.632415   
4887                       2.0          -0.100096           0.134536   

      DurationOfPitch_std  TypeofContact_label  ...  \
0               -1.164769                    1  ...   
1               -0.166872                    0  ...   
2               -0.915295                    1  ...   
3               -0.790558                    0  ...   
4               -0.915295                    1  ...   
...                   ...                  ...  ...   
4883            -0.790558                    1  ...   
4884             1.953659                    0  ...   
4885             0.207339                    1  ...   
4886             0.082602                    1  ...   
4887            -0.166872                    1  ...   

      ProductPitched_Super Deluxe  MaritalStatus_Divorced  \
0                               0                       0   
1                               0                       1   
2                               0                       0   
3                               0                       1   
4                               0                       1   
...                           ...                     ...   
4883                            0                       0   
4884                            0                       0   
4885                            0                       0   
4886                            0                       0   
4887                            0                       0   

      MaritalStatus_Married  MaritalStatus_Single  MaritalStatus_Unmarried  \
0                         0                     1                        0   
1                         0                     0                        0   
2                         0                     1                        0   
3                         0                     0                        0   
4                         0                     0                        0   
...                     ...                   ...                      ...   
4883                      0                     0                        1   
4884                      0                     1                        0   
4885                      1                     0                        0   
4886              

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4787 entries, 0 to 4887
Data columns (total 34 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   CustomerID                    4787 non-null   int64  
 1   ProdTaken                     4787 non-null   object 
 2   Age                           4787 non-null   float64
 3   NumberOfPersonVisiting        4787 non-null   int64  
 4   NumberOfFollowups             4787 non-null   float64
 5   NumberOfChildrenVisiting      4787 non-null   float64
 6   NumberOfTrips_std             4787 non-null   float64
 7   MonthlyIncome_std             4787 non-null   float64
 8   DurationOfPitch_std           4787 non-null   float64
 9   TypeofContact_label           4787 non-null   int8   
 10  Passport_label                4787 non-null   int8   
 11  OwnCar_label                  4787 non-null   int8   
 12  Gender_label                  4787 non-null   int8   
 13  Pre

In [30]:
# df.to_csv('Data.csv', index = False)

## Drop Fitur Menggunakan VIF

https://towardsdatascience.com/how-to-remove-multicollinearity-using-python-4da8d9d8abb2

In [31]:
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor

X = df[list(df.columns[2:])]
vif_info = pd.DataFrame()
vif_info['VIF'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
vif_info['Column'] = X.columns
vif_info.sort_values('VIF', ascending=False)

C:\Users\ASUS\anaconda3\lib\site-packages\statsmodels\stats\outliers_influence.py:193: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)


VIF                        Column
16       inf           Occupation_Salaried
17       inf     Occupation_Small Business
30       inf    Designation_Senior Manager
29       inf           Designation_Manager
28       inf         Designation_Executive
27       inf               Designation_AVP
26       inf       MaritalStatus_Unmarried
25       inf          MaritalStatus_Single
24       inf         MaritalStatus_Married
23       inf        MaritalStatus_Divorced
22       inf   ProductPitched_Super Deluxe
21       inf       ProductPitched_Standard
20       inf           ProductPitched_King
19       inf         ProductPitched_Deluxe
18       inf          ProductPitched_Basic
31       inf                Designation_VP
15       inf     Occupation_Large Business
14       inf        Occupation_Free Lancer
5   5.631542             MonthlyIncome_std
1   1.836591        NumberOfPersonVisiting
3   1.684509      NumberOfChildrenVisiting
0   1.414847                           Age
2   1.215427             NumberOfFollowups
13  1.131299                CityTier_label
4   1.096422             NumberOfTrips_std
6   1.024716           DurationOfPitch_std
9   1.016169                  OwnCar_label
12  1.014694  PitchSatisfactionScore_label
11  1.012121   PreferredPropertyStar_label
10  1.011564                  Gender_label
8   1.009008                Passport_label
7   1.008202           TypeofContact_label

Fitur yang akan digunakan untuk modeling adalah fitur dengan VIF < 10 yaitu : MonthlyIncome_std, NumberOfPersonVisiting, NumberOfChildrenVisiting, Age, NumberOfFollowups, CityTier_label, NumberOfTrips_norm, DurationOfPitch_std, PitchSatisfactionScore_label, OwnCar_label, Gender_label, PreferredPropertyStar_label, Passport_label, dan TypeofContact_label.

Berdasarkan artikel di bawah, VIF dengan nilai inf menunjukkan bahwa terdapat korelasi yang sempurna

https://www.sigmamagic.com/blogs/what-is-variance-inflation-factor/#:~:text=If%20there%20is%20perfect%20correlation,a%20correlation%20between%20the%20variables.&text=A%20general%20rule%20of%20thumb,10%20then%20there%20is%20multicollinearity.

## Data Pre-Processing Conclusion

- Data awal : 4888 rows x 20 columns
- Data setelah pre-processing (sampai encoding): 4787 rows x 34 columns (masih ada kolom CustomerID)